docs/architecture/adr-039-add-support-for-wasm-light-clients.md

```diff
@@ -0,0 +1,107 @@
+event of security issue or consensus update, we need to modify the codebase and integrate it in numerous places.
+
+2. Governance voting: Adding new light client implementations require governance support and is expensive: This is
+not ideal as chain governance is gatekeeper for new light client implementations getting added. If a small community
+want support for light client X, they may not be able to convince governance to support it.
+
+3. Validator upgrade: After governance voting succeeds, validators need to upgrade their nodes in order to enable new
+IBC light client implementation. This is both time consuming and error prone.
+   
+Another problem stemming from above process is, if a chain want to upgrade its own consensus, it will need to convince every chain
+or hub connected to it to upgrade its light client in order to stay connected. Due to time consuming process required
+to upgrade light client, a chain with lots of connection need to be disconnected for quite some time after upgrading 
+its consensus, which can be very expensive.
+
+We are proposing simplifying this workflow by integrating a WASM light client module which makes adding support for
+a new light client a simple transaction. The light client bytecode, written in Wasm-compilable Rust, runs inside a Wasmer
+VM. The Wasm light client submodule exposes a proxy light client interface that routes incoming messages to the
+appropriate handler function, inside the Wasm VM for execution.
+
+With WASM light client module, anybody can add new IBC light client in the form of WASM bytecode (provided they are able to pay huge gas fee for the transaction)
+as well as instantiate clients using any created client type. This allows any chain to update its own light client in other chains
+without going through steps outlined above.
+
+
+## Decision
+
+We decided to use WASM light client module as a generic light client which will interface with the actual light client
+uploaded as WASM bytecode. This will require changing client selection method to allow any client if the client type
+has prefix of `wasm/`.
+
+```go
+// IsAllowedClient checks if the given client type is registered on the allowlist.
+func (p Params) IsAllowedClient(clientType string) bool {
+	for _, allowedClient := range p.AllowedClients {
+		if allowedClient == clientType {
+			return true
+		}
+	}
+	return false || (p.AreWASMClientsAllowed && isWASMClient(clientType))
+}
+```
+
+Inside Wasm light client `ClientState`, appropriate Wasm bytecode will be executed depending upon `ClientType`.
+
+```go
+func (cs ClientState) Validate() error {
+    wasmRegistry = getWASMRegistry()
+	  clientType := cs.ClientType()
+    codeHandle := wasmRegistry.getCodeHandle(clientType)
+    return codeHandle.validate(cs)
+}
+```
+
+To upload new light client, user need to create a transaction with Wasm byte code which will be
+processed by IBC Wasm module.
```

Comment:
So the WASM module will need to add messages and a keeper?

Reply:
@colin-axner Yes, a keeper is needed to keep track of different versions uploaded as well as add support for querying latest version, get binary blob of latest wasm code etc etc.

Reply:
Is there a way we could modify the existing implementation to allow light client implementations to make use of the client prefixed store? It seems like a flaw in the existing design if you need to create a second store

Reply:
@colin-axner The existing client prefixed store is used to store client state and related data structure and this store is being passed to wasm vm, the second store is for book-keeping purpose where the versioning info and management info is stored.